content/blog/2023/2023-01-25-junit5-parameterized-tests.md

```diff
@@ -0,0 +1,460 @@
+By default, each argument provided to a `@ParameterizedTest` method corresponds to a `single method parameter`. Due to this, when argument sources that supply a large number of arguments can lead to large method signatures.
+To solve this problem, we can use `ArgumentsAccessor` instead of declaring multiple parameters. The type conversion is supported as discussed in Implicit conversion above.
+
+```java
+@ParameterizedTest
+@CsvSource({ "John, 20", 
+		     "Harry, 30" })
+void checkArgumentsAccessor(ArgumentsAccessor arguments) {
+	Person person = new Person(arguments.getString(0), arguments.getInteger(1));
+	assertTrue(person.getAge() > 19, person.getName() + " is a teenager");
+}
+```
+
+#### Custom Aggregators
+
+We saw using an `ArgumentsAccessor` can access the @ParameterizedTest method’s arguments directly. What if we want to declare the same ArgumentsAccessor in multiple tests? JUnit5 solves this by providing custom, reusable aggregators.
+
+- **@AggregateWith** - Implement the `ArgumentsAggregator` interface and register it via the `@AggregateWith` annotation in the @ParameterizedTest method. When we execute the test, it provides the aggregation result as an argument for the corresponding test. The implementation of ArgumentsAggregator can be an external class or a static nested class.
+
+```java
+@ParameterizedTest
+@CsvSource({ "John, 20", 
+			 "Harry, 30" })
+void checkArgumentsAggregator(@AggregateWith(PersonArgumentsAggregator.class) Person person) {
+	assertTrue(person.getAge() > 19, person.getName() + " is a teenager");
+}
+
+public class PersonArgumentsAggregator implements ArgumentsAggregator {
+
+	@Override
+	public Object aggregateArguments(ArgumentsAccessor arguments, ParameterContext context)
+			throws ArgumentsAggregationException {
+		return new Person(arguments.getString(0), arguments.getInteger(1));
+	}
+}
+```
+
+- **Custom Annotation** - When we see we are repeating `@AggregateWith(PersonArgumentsAggregator.class)` declaration for each test, we can create a custom annotation such as `@CsvToPerson` that can be meta-annotated with `@AggregateWith(PersonArgumentsAggregator.class)` to have a similar effect.
+
+```java
+@ParameterizedTest
+@CsvSource({ "John, 20", 
+             "Harry, 30" })
+void checkCustomAggregatorAnnotation(@CsvToPerson Person person) {
+	assertTrue(person.getAge() > 19, person.getName() + " is a teenager");
+}
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+@AggregateWith(PersonArgumentsAggregator.class)
+public @interface CsvToPerson {
+}
+```
+
+## Summary
+
+JUnit5 parameterized tests help us avoid duplicate tests and provide the ability to execute the same test several times using different inputs.
+
+Since you have read the article to the end, I would like to give you a bonus - If you're using assertion frameworks like - [Fluent assertions for java](https://joel-costigliola.github.io/assertj/) you can pass the Consumer as an argument that holds the assertion itself.
```

Comment:
Please remove this section from the summary and place it in the main body of the article.The summary should provide a brief overview of the main points covered in the article.

Reply:
Moved.